pysb/simulator/scipyode.py

```diff
@@ -384,3 +410,57 @@ def run(self, tspan=None, initials=None, param_values=None):
         tout = np.array([self.tspan]*n_sims)
         self._logger.info('All simulation(s) complete')
         return SimulationResult(self, tout, trajectories)
+
+
+@contextlib.contextmanager
+def _patch_distutils_logging(base_logger):
+    """Patch distutils logging functionality with logging.Logger calls.
+
+    The value of the 'base_logger' argument should be a logging.Logger instance,
+    and its effective level will be passed on to the patched distutils loggers.
+
+    distutils.log contains its own internal PEP 282 style logging system that
+    sends messages straight to stdout/stderr, and numpy.distutils.log extends
+    that. This code patches all of this with calls to logging.LoggerAdapter
+    instances, and disables the module-level threshold-setting functions so we
+    can retain full control over the threshold. Also all WARNING messages are
+    "downgraded" to INFO to suppress excessive use of WARNING-level logging in
+    numpy.distutils.
+
+    """
+    logger = get_logger(__name__)
+    logger.debug('patching distutils and numpy.distutils logging')
+    logger_methods = 'log', 'debug', 'info', 'warn', 'error', 'fatal'
+    other_functions = 'set_threshold', 'set_verbosity'
+    saved_symbols = {}
+    for module_name in 'distutils.log', 'numpy.distutils.log':
+        new_logger = _DistutilsProxyLoggerAdapter(
+            base_logger, {'module': module_name}
+        )
+        module = importlib.import_module(module_name)
+        # Save the old values.
+        for name in logger_methods + other_functions:
+            saved_symbols[module, name] = getattr(module, name)
+        # Replace logging functions with bound methods of the Logger object.
+        for name in logger_methods:
+            setattr(module, name, getattr(new_logger, name))
+        # Replace threshold-setting functions with no-ops.
+        for name in other_functions:
+            setattr(module, name, lambda *args, **kwargs: None)
+    try:
+        yield
+    finally:
+        logger.debug('restoring distutils and numpy.distutils logging')
+        # Restore everything we overwrote.
+        for (module, name), value in saved_symbols.items():
+            setattr(module, name, value)
+
+
+class _DistutilsProxyLoggerAdapter(logging.LoggerAdapter):
+    """A logging adapter for the distutils logging patcher."""
+    def process(self, msg, kwargs):
+        return '(from %s) %s' % (self.extra['module'], msg), kwargs
```

Comment:
This ends up logging the distutils.log messages under pysb.simulator.scipyode, with a note about which distutils module the call actually originated from. Should we instead create Loggers for those modules directly, which will require some extra logger setup code on our part?

Reply:
I think the current approach is sound. It's more useful to know the message came through a Simulator than just seeing that it came from distutils.